source/core/index.ts

```diff
@@ -0,0 +1,1472 @@
+	lookup?: CacheableLookup['lookup'];
+	localAddress?: string;
+	createConnection?: Options['createConnection'];
+}
+
+export interface Progress {
+	percent: number;
+	transferred: number;
+	total?: number;
+}
+
+export interface PlainResponse extends IncomingMessageWithTimings {
+	requestUrl: string;
+	redirectUrls: string[];
+	request: Request;
+	ip?: string;
+	isFromCache: boolean;
+	statusCode: number;
+	url: string;
+	timings: Timings;
+}
+
+// For Promise support
+export interface Response<T = unknown> extends PlainResponse {
+	body: T;
+	retryCount: number;
+}
+
+export interface RequestEvents<T> {
+	on(name: 'request', listener: (request: http.ClientRequest) => void): T;
+	on(name: 'response', listener: (response: Response) => void): T;
+	on(name: 'redirect', listener: (response: Response, nextOptions: NormalizedOptions) => void): T;
+	on(name: 'uploadProgress' | 'downloadProgress', listener: (progress: Progress) => void): T;
+}
+
+function validateSearchParams(searchParams: Record<string, unknown>): asserts searchParams is Record<string, string | number | boolean | null> {
+	// eslint-disable-next-line guard-for-in
+	for (const key in searchParams) {
+		const value = searchParams[key];
+
+		if (!is.string(value) && !is.number(value) && !is.boolean(value) && !is.null_(value)) {
+			throw new TypeError(`The \`searchParams\` value '${String(value)}' must be a string, number, boolean or null`);
+		}
+	}
+}
+
+function isClientRequest(clientRequest: unknown): clientRequest is ClientRequest {
+	return is.object(clientRequest) && !('statusCode' in clientRequest);
+}
+
+const cacheFn = async (url: URL, options: RequestOptions): Promise<ClientRequest | ResponseLike> => new Promise<ClientRequest | ResponseLike>((resolve, reject) => {
+	// TODO: Remove `utils/url-to-options.ts` when `cacheable-request` is fixed
+	Object.assign(options, urlToOptions(url));
+
+	// `http-cache-semantics` checks this
+	delete (options as unknown as NormalizedOptions).url;
+
+	// TODO: `cacheable-request` is incorrectly typed
+	const cacheRequest = (options as Pick<NormalizedOptions, 'cacheableRequest'>).cacheableRequest!(options, resolve as any);
+
+	// Restore options
+	(options as unknown as NormalizedOptions).url = url;
+
+	cacheRequest.once('error', (error: Error) => {
+		if (error instanceof CacheableRequest.RequestError) {
+			// TODO: `options` should be `normalizedOptions`
+			reject(new RequestError(error.message, error, options as unknown as NormalizedOptions));
+			return;
+		}
+
+		// TODO: `options` should be `normalizedOptions`
+		reject(new CacheError(error, options as unknown as NormalizedOptions));
+	});
+	cacheRequest.once('request', resolve);
+});
+
+const waitForOpenFile = async (file: ReadStream): Promise<void> => new Promise((resolve, reject) => {
+	const onError = (error: Error): void => {
+		reject(error);
+	};
+
+	file.once('error', onError);
+	file.once('open', () => {
+		file.off('error', onError);
+		resolve();
+	});
+});
```

Comment:
This is just 275 bytes long, there's no need to include `p-event` (34.1kB) in the dependencies.